mac80211: ath11k: add support for dynamic vlan (ap/vlan) patch

davidrapan opened the pull request:
Add AP/VLAN supp for ath11k. Long-term tested, including active use of VLANs according to passwords + FT.
* [linux-wireless.vger.kernel.org archive mirror: [PATCH v3] ath11k: add support for dynamic vlan (ap/vlan)](https://lore.kernel.org/linux-wireless/DB6PR08MB2693100915904FB05300CA47DF99A@DB6PR08MB2693.eurprd08.prod.outlook.com/)

Tested on AX3600 (qualcommax, ipq807x):
```
root@starbase:~# iw list
...
Wiphy phy0
        wiphy index: 0
        max # scan SSIDs: 16
        max scan IEs length: 142 bytes
        max # sched scan SSIDs: 0
        max # match sets: 0
        Retry short limit: 7
        Retry long limit: 4
        Coverage class: 0 (up to 0m)
        Device supports AP-side u-APSD.
        Available Antennas: TX 0xf0 RX 0xf0
        Configured Antennas: TX 0xf0 RX 0xf0
        Supported interface modes:
                 * managed
                 * AP
                 * AP/VLAN
                 * monitor
                 * mesh point
        Band 2:
                Capabilities: 0x19e7
                        RX LDPC
                        HT20/HT40
                        Dynamic SM Power Save
                        RX HT20 SGI
                        RX HT40 SGI
                        TX STBC
                        RX STBC 1-stream
                        Max AMSDU length: 7935 bytes
                        DSSS/CCK HT40
                Maximum RX AMPDU length 65535 bytes (exponent: 0x003)
                Minimum RX AMPDU time spacing: No restriction (0x00)
                HT TX/RX MCS rate indexes supported: 0-31
                VHT Capabilities (0x339b79fa):
                        Max MPDU length: 11454
                        Supported Channel Width: 160 MHz, 80+80 MHz
                        RX LDPC
                        short GI (80 MHz)
                        short GI (160/80+80 MHz)
                        TX STBC
                        SU Beamformer
                        SU Beamformee
                        MU Beamformer
                        MU Beamformee
                        RX antenna pattern consistency
                        TX antenna pattern consistency
...
```

I believe this is an updated version compared to: https://github.com/openwrt/openwrt/pull/17644

./David

robimarko commented:
This should go upstream

davidrapan commented:
I think so too, but there was an attempt here once in 2023 and for some reason it didn't work out (the PR probably rotted away like many others that no one found the time for), so maybe we can at least start somewhere.

robimarko commented:
It rotted out exactly because I do not want to carry this locally forever, it should be first upstreamed to ath11k

davidrapan commented:
I was talking about submit to the upstream, which rotted, twice...

I'll try to get it upstreamed, but I'm not getting my hopes up.

gtxaspec commented:
i know that upstreaming is the ideal path forward, but given the proven functionality and on device testing, i think it makes practical sense to carry this patch locally in the interim. users shouldn't have to wait indefinitely for a core networking feature that's already working well on production hardware. i've been testing this for some time on linksys mx4200/mx4300 units with proven success as well. for an openwrt platform as popular as ath11k is now, having vlan feature parity with other chipsets seems worth the temporary maintenance cost.

**continues waiting and crossing fingers**

robimarko commented:
@gtxaspec Well, that is the thing, its not going to be temporary, its just going to be kept forever downstream.
I dont want to keep having to update old patches that were never upstreamed.

davidrapan commented:
Here you go [linux-wireless.vger.kernel.org archive mirror: [PATCH] ath11k: add support for dynamic vlan (ap/vlan)](https://lore.kernel.org/linux-wireless/DB6PR08MB2693ADF1B27329C7BCED7AA5DF92A@DB6PR08MB2693.eurprd08.prod.outlook.com/)

castiel652 commented:
> Here you go [linux-wireless.vger.kernel.org archive mirror: [PATCH] ath11k: add support for dynamic vlan (ap/vlan)](https://lore.kernel.org/linux-wireless/DB6PR08MB2693ADF1B27329C7BCED7AA5DF92A@DB6PR08MB2693.eurprd08.prod.outlook.com/)

You should split it into two patches. One for ath11k and the other one for mac80211.

castiel652 commented on package/kernel/mac80211/patches/ath11k/948-wifi-ath11k-add-support-for-dynamic-vlan.patch:
```diff
@@ -0,0 +1,381 @@
+ 	case HAL_TCL_ENCAP_TYPE_NATIVE_WIFI:
+-		ath11k_dp_tx_encap_nwifi(skb);
++		if (arvif->vif->offload_flags & IEEE80211_OFFLOAD_ENCAP_ENABLED)
++			is_diff_encap = true;
```
This will break management frame under 802.3 mode.

castiel652 commented on package/kernel/mac80211/patches/ath11k/948-wifi-ath11k-add-support-for-dynamic-vlan.patch:
```diff
@@ -0,0 +1,381 @@
+ 	case HAL_TCL_ENCAP_TYPE_NATIVE_WIFI:
+-		ath11k_dp_tx_encap_nwifi(skb);
++		if (arvif->vif->offload_flags & IEEE80211_OFFLOAD_ENCAP_ENABLED)
++			is_diff_encap = true;
```
This will break management frame under 802.3 mode.

davidrapan replied:
Not sure what do you mean, what exactly shouldn't work?

castiel652 replied:
From what I remember, management frames are sent in `Native WiFi` even when vif is set to encap offload mode.

castiel652 replied:
Nvm. I got confused. management frames actually don't go through this path.
But why does it need to skip encap when offload is enabled?

castiel652 replied:
But it doesn't skip encap in the original patch, ath12k doesn't either.
Based on the description that part is only for the firmware.

davidrapan replied:
I know, there's a lot more going on than just this, and the latest revision from the original author skips encap too.

castiel652 replied:
I see.
Then I would suggest makes it actually go through `HAL_TCL_ENCAP_TYPE_RAW` instead of using `is_diff_encap`.
There will be extra work but I think it's cleaner that way.
You can see ath10k for reference.
https://patchwork.kernel.org/project/linux-wireless/patch/1542798288-8946-1-git-send-email-mpubbise@codeaurora.org/

castiel652 commented on package/kernel/mac80211/patches/subsys/948-wifi-mac80211-add-support-for-dynamic-vlan.patch:
```diff
@@ -0,0 +1,61 @@
++	if (sdata->vif.type == NL80211_IFTYPE_AP_VLAN) {
++		ap_sdata = container_of(sdata->bss,
++				 struct ieee80211_sub_if_data, u.ap);
++		if (ap_sdata->vif.offload_flags & IEEE80211_OFFLOAD_ENCAP_ENABLED &&
```
How do you have this flag set with AP_VLAN?
Neither mac80211 or ath11k allows that.

davidrapan replied:
When I was originally testing the code, like a year ago I was debugging the code and they were both set:
```
kern.debug kernel: [1368.113582] phy0-ap0: sdata->vif.offload_flags: 3
```

castiel652 replied:
> When I was originally testing the code, like a year ago I was debugging the code and they were both set:
> 
> ```
> kern.debug kernel: [1368.113582] phy0-ap0: sdata->vif.offload_flags: 3
> ```

My apologies. You are correct.
But then there's another problem.
If `IEEE80211_OFFLOAD_ENCAP_ENABLED` is set, it uses `ieee80211_subif_start_xmit_8023`.
Does the packet fall through the slow path?

castiel652 replied:
Okay so it does fall through but I think it should meet the check in fast path instead of fall to slow path then jump back.
Also I found this https://patchwork.kernel.org/project/linux-wireless/patch/20240506212014.670423-3-quic_msinada@quicinc.com/

castiel652 replied:
I have a question since I am not familiar with AP_VLAN.
Does the vlan unicast frame come from the `ieee80211_convert_to_unicast`?

castiel652 replied:
Thanks for the explanation.
I think it got rejected because it looks hacky.

castiel652 replied:
I saw that. :)
You didn't split it into two patches.
It should be two patches like this
1. ath11k: add support for dynamic vlan
2. mac80211: add support for vlan unicast packet for 802.3 xmit path

castiel652 replied:
Yes they are all in wirless mailing list but they have different maintainers/reviewers.
ath11k patches are reviewed by Jeff Johnson and mac80211 patches are reviewed by Johannes Berg.
That's why I said you should split it into two patches.

davidrapan replied:
Argh, it's just so tedious method of contribution. 😞

castiel652 replied:
That's just how it works if you want your patch accepted

davidrapan replied:
> how it works

I know that. What I don't know (or rather I'm very confused of) is ath11k lack of support for dynamic vlan (on top of that w/o at least SOME pressure to implement it)...

castiel652 replied:
Qualcomm usually stops resending patches unless it's a crucial feature or bugfix.
so a lot of time it's us community that does the work :P

davidrapan committed:
mac80211: ath11k: add support for dynamic vlan

+ software encryption of VLAN group traffics

Tested-on: IPQ8074 hw2.0 AHB WLAN.HK.2.9.0.1-01385-QCAHKSWPL_SILICONZ-1
Tested-on: IPQ8074 hw2.0 AHB WLAN.HK.2.9.0.1-01977-QCAHKSWPL_SILICONZ-1

Signed-off-by: David Rapan <david@rapan.cz>

gtxaspec commented:
@davidrapan any progress on any fronts?
